### PR TITLE
Fixes #29860 - Add base macros docs

### PR DESCRIPTION
**Pull request by ofedoren**
Added docs for basic macros, new sections and `host_[memory | sockets | cores | virtual]` methods to allowed.

### CODE REVIEW

File: lib/foreman/renderer/scope/macros/base.rb [hunk 1]
```diff
@@ -11,36 +12,85 @@ module Base
...
             subnet.parameters.exists?(name: param_name)
           end
 
+          apipie :method, 'Returns the value of global setting' do
+            required :name, String, desc: 'the name of the setting which can be found by hovering the setting with mouse cursor in the UI, or via API/CLI'
+            optional :blank_default, Object, desc: 'if the setting is not set to any value, this value will be returned instead', default: nil
+            returns Object, desc: 'The value of global setting, e.g. String, Integer, Array, Provisioning Template etc'
+            example "global_setting('foreman_url') # => 'https://foreman.example.com'"
+            example "global_setting('outofsync_interval', 30) # => 30"
+          end
```
Comment (lzap) [inline lib/foreman/renderer/scope/macros/base.rb:41]: Actually an important fact is that not all settings are exposed, only those which are allowed via `Foreman::Renderer.config.allowed_global_settings`.

File: lib/foreman/renderer/scope/macros/base.rb [hunk 2]
```diff
@@ -49,10 +99,28 @@ def pxe_kernel_options
             ''
           end
 
+          apipie :method, 'Generates a shell command to store the given text into a file' do
+            desc "This is useful if some multline string needs to be saved somewhere on the harddisk. This
```
Comment (lzap) [inline lib/foreman/renderer/scope/macros/base.rb]: multiline

File: lib/foreman/renderer/scope/macros/base.rb [hunk 3]
```diff
@@ -49,10 +99,28 @@ def pxe_kernel_options
...
+              is typically used in provisioning or job templates, e.g. when puppet configuration file is
+              generated based on host configuration and stored for puppet agent."
+            required :filename, String, desc: 'the file path to store the content to'
+            required :content, String, desc: 'content to be stored'
+            returns String, desc: 'String representing the shell command'
+            example "save_to_file(\"hello\\nworld\\n\", '/etc/motd') # => 'cat << EOF > /etc/motd\\nhello\\nworld\\nEOF'"
+          end
           def save_to_file(filename, content)
             "cat << EOF > #{filename}\n#{content}EOF"
```
Comment (lzap) [inline lib/foreman/renderer/scope/macros/base.rb:114]: Not part of the patch, but I'd probably avoid the generic "EOF" which people use a lot.

File: lib/foreman/renderer/scope/macros/base.rb [hunk 4]
```diff
@@ -91,64 +177,156 @@ def default_template_url(template, hostgroup)
...
+            required :mac_address, String, desc: 'mac address in a format of hexadecimal numbers separated by colon'
+            returns String, desc: 'A name that can be used as a hostname. It is based on passed mac-address, same
+              mac-address results in the same hostname.'
+            example 'mac_name("00:11:22:33:44:55") # => "hazel-diana-maruyama-feltus"'
+          end
           def mac_name(mac_address)
             NameGenerator.new.generate_next_mac_name(mac_address)
           end
 
+          apipie :method, 'Returns a kernel release installed on a host based on facts reported by facters' do
```
Comment (lzap) [inline lib/foreman/renderer/scope/macros/base.rb]: by facter

File: lib/foreman/renderer/scope/macros/base.rb [hunk 5]
```diff
@@ -91,64 +177,156 @@ def default_template_url(template, hostgroup)
...
+            required :host, 'Host::Managed', desc: 'host object for which kernel released is returned'
+            returns String, desc: 'String describing the kernel release'
+            example 'host_kernel_release(host) # => "3.10.0-957.10.1.el7.x86_6"'
+            example "<%- load_hosts(preload: :kernel_release).each_record do |host| -%>\n<%=   host.name -%>, <%=  host_kernel_release(host) %>\n<%- end -%>"
+          end
           def host_kernel_release(host)
             host&.kernel_release&.value
           end
 
+          apipie :method, 'Returns a host uptime in seconds based on facts reported by facters' do
```
Comment (lzap) [inline lib/foreman/renderer/scope/macros/base.rb]: ditto, actually you can drop "reported by facters". It's just a fact, it can be reported by Ansible too.

File: lib/foreman/renderer/scope/macros/base.rb [hunk 6]
```diff
@@ -91,64 +177,156 @@ def default_template_url(template, hostgroup)
...
+          apipie :method, 'Returns true if the given host is virtual, false otherwise' do
+            required :host, 'Host::Managed', desc: "host object to check if it's virtual"
+            returns one_of: [true, false]
+            example 'host_virtual(host) # => true'
+          end
           def host_virtual(host)
             host&.virtual
           end
 
+          apipie :method, 'Compares two gem versions' do
```
Comment (lzap) [inline lib/foreman/renderer/scope/macros/base.rb]: It can be used to compare RPM or DEB package versions but it will not handle edge cases (e.g. 1.0.0.pre1-snapshot-1).

**Comment by ofedoren** [comment]
@lzap, thank you for review! Updated.

File: lib/foreman/renderer/scope/macros/base.rb (hunk 1 above)
Comment (lzap) [reply lib/foreman/renderer/scope/macros/base.rb:41]: Can't you simply list them in the documentation? It is just an array I think, append it.

**Comment by lzap** [comment]
Rebase needed

**Comment by ofedoren** [comment]
@lzap, rebased.

File: lib/foreman/renderer/scope/macros/base.rb (hunk 1 above)
Comment (ofedoren) [reply lib/foreman/renderer/scope/macros/base.rb:41]: @lzap, sorry didn't notice your comment. Added list of the allowed settings.

File: lib/foreman/renderer/scope/macros/base.rb [hunk 7]
```diff
@@ -91,64 +179,157 @@ def default_template_url(template, hostgroup)
...
+            required :host, 'Host::Managed', desc: "host object to check if it's virtual"
+            returns one_of: [true, false]
+            example 'host_virtual(host) # => true'
+          end
           def host_virtual(host)
             host&.virtual
           end
 
+          apipie :method, 'Compares two package versions' do
+            desc 'It can be used to compare RPM, DEB or gem package versions'
```
Comment (ares) [inline lib/foreman/renderer/scope/macros/base.rb]: lzap asked for changing this to "It can be used to compare RPM or DEB package versions but it will not handle edge cases (e.g. 1.0.0.pre1-snapshot-1)."

**Review by ares — CHANGES_REQUESTED** [review]
one review comment missed, texts are imho otherwise ok, we could improve formatting here and there butt let's get it in and improve iteratively

**Comment by ofedoren** [comment]
@ares, you're right, thanks. Updated.

**Review by ares — APPROVED** [review]
Thanks @ofedoren, merging!
